File: java/exercises/practice/unit_test_072f9e66_91a8_478b_9237_f44fc7501470/task_scheduler/.meta/src/main/java/TaskScheduler.java
import java.util.*;

public class TaskScheduler {
    
    /**
     * Schedules tasks optimally to maximize the number completed before their deadlines.
     * 
     * @param tasks List of tasks to be scheduled
     * @param systemResources Available system resources
     * @return Ordered list of task IDs representing the optimal execution schedule
     */
    public List<String> schedule(List<Task> tasks, Map<String, Integer> systemResources) {
        if (tasks.isEmpty()) {
            return new ArrayList<>();
        }
        
        // Create a graph representation for dependency checking
        Map<String, Set<String>> dependencyGraph = buildDependencyGraph(tasks);
        Map<String, Task> taskMap = createTaskMap(tasks);
        
        // Check for circular dependencies
        if (hasCircularDependencies(dependencyGraph, taskMap)) {
            return new ArrayList<>();
        }
        
        // This will track the completion time for each task
        Map<String, Integer> completionTimes = new HashMap<>();
        
        // Keep track of available resources at each point in time
        Map<String, Integer> availableResources = new HashMap<>(systemResources);
        
        // Keeps track of tasks that have been scheduled
        Set<String> scheduledTasks = new HashSet<>();
        
        // The resulting schedule
        List<String> schedule = new ArrayList<>();
        
        // Current time in the simulation
        int currentTime = 0;
        
        // Loop until all tasks are scheduled or no more tasks can be scheduled
        while (scheduledTasks.size() < tasks.size()) {
            List<Task> eligibleTasks = new ArrayList<>();
            
            // Find all eligible tasks at the current time
            for (Task task : tasks) {
                if (!scheduledTasks.contains(task.getId()) && 
                    areAllDependenciesMet(task, scheduledTasks) && 
                    enoughResourcesAvailable(task, availableResources)) {
                    
                    // Check if the task can finish before its deadline
                    if (currentTime + task.getExecutionTime() <= task.getDeadline()) {
                        eligibleTasks.add(task);
                    }
                }
            }
            
            if (eligibleTasks.isEmpty()) {
                // If no eligible tasks, try to advance time to next possible task
                boolean advanced = advanceTimeToNextEligibleTask(
                    currentTime, tasks, scheduledTasks, completionTimes, 
                    availableResources, systemResources);
                
                if (!advanced) {
                    // No tasks can be scheduled anymore
                    break;
                }
            } else {
                // Select the best task based on priority heuristic
                Task selectedTask = selectBestTask(eligibleTasks);
                
                // Schedule the selected task
                schedule.add(selectedTask.getId());
                scheduledTasks.add(selectedTask.getId());
                
                // Update resources
                allocateResources(selectedTask, availableResources);
                
                // Update time and completion time for this task
                int taskCompletionTime = currentTime + selectedTask.getExecutionTime();
                completionTimes.put(selectedTask.getId(), taskCompletionTime);
                currentTime = taskCompletionTime;
                
                // Release resources for completed tasks
                releaseResources(currentTime, completionTimes, taskMap, availableResources);
            }
        }
        
        // Final verification - ensure the schedule is valid
        if (!isValidSchedule(schedule, taskMap, systemResources)) {
            return new ArrayList<>();
        }
        
        return schedule;
    }
    
    /**
     * Builds a graph representation of task dependencies.
     */
    private Map<String, Set<String>> buildDependencyGraph(List<Task> tasks) {
        Map<String, Set<String>> graph = new HashMap<>();
        
        for (Task task : tasks) {
            graph.put(task.getId(), new HashSet<>());
        }
        
        for (Task task : tasks) {
            for (String dependency : task.getDependencies()) {
                graph.get(dependency).add(task.getId());
            }
        }
        
        return graph;
    }
    
    /**
     * Creates a map of task IDs to Task objects for quick lookup.
     */
    private Map<String, Task> createTaskMap(List<Task> tasks) {
        Map<String, Task> taskMap = new HashMap<>();
        for (Task task : tasks) {
            taskMap.put(task.getId(), task);
        }
        return taskMap;
    }
    
    /**
     * Checks if the dependency graph has circular dependencies using DFS.
     */
    private boolean hasCircularDependencies(Map<String, Set<String>> graph, Map<String, Task> taskMap) {
        Set<String> visited = new HashSet<>();
        Set<String> recursionStack = new HashSet<>();
        
        for (String taskId : graph.keySet()) {
            if (isCircular(taskId, graph, visited, recursionStack, taskMap)) {
                return true;
            }
        }
        
        return false;
    }
    
    /**
     * Helper method for detecting cycles in the dependency graph.
     */
    private boolean isCircular(String taskId, Map<String, Set<String>> graph, 
                              Set<String> visited, Set<String> recursionStack,
                              Map<String, Task> taskMap) {
        if (recursionStack.contains(taskId)) {
            return true;
        }
        
        if (visited.contains(taskId)) {
            return false;
        }
        
        visited.add(taskId);
        recursionStack.add(taskId);
        
        // Check all dependencies of the current task
        for (String dependency : taskMap.get(taskId).getDependencies()) {
            if (graph.containsKey(dependency) && 
                isCircular(dependency, graph, visited, recursionStack, taskMap)) {
                return true;
            }
        }
        
        recursionStack.remove(taskId);
        return false;
    }
    
    /**
     * Checks if all dependencies of a task have been scheduled.
     */
    private boolean areAllDependenciesMet(Task task, Set<String> scheduledTasks) {
        for (String dependency : task.getDependencies()) {
            if (!scheduledTasks.contains(dependency)) {
                return false;
            }
        }
        return true;
    }
    
    /**
     * Checks if enough resources are available to execute a task.
     */
    private boolean enoughResourcesAvailable(Task task, Map<String, Integer> availableResources) {
        for (Map.Entry<String, Integer> entry : task.getResourceRequirements().entrySet()) {
            String resourceType = entry.getKey();
            int requiredAmount = entry.getValue();
            
            if (!availableResources.containsKey(resourceType) || 
                availableResources.get(resourceType) < requiredAmount) {
                return false;
            }
        }
        return true;
    }
    
    /**
     * Allocates resources for a task execution.
     */
    private void allocateResources(Task task, Map<String, Integer> availableResources) {
        for (Map.Entry<String, Integer> entry : task.getResourceRequirements().entrySet()) {
            String resourceType = entry.getKey();
            int requiredAmount = entry.getValue();
            
            availableResources.put(resourceType, availableResources.get(resourceType) - requiredAmount);
        }
    }
    
    /**
     * Releases resources from completed tasks.
     */
    private void releaseResources(int currentTime, Map<String, Integer> completionTimes, 
                                 Map<String, Task> taskMap, Map<String, Integer> availableResources) {
        for (Map.Entry<String, Integer> entry : completionTimes.entrySet()) {
            String taskId = entry.getKey();
            int completionTime = entry.getValue();
            
            if (completionTime <= currentTime) {
                Task task = taskMap.get(taskId);
                
                for (Map.Entry<String, Integer> resourceEntry : task.getResourceRequirements().entrySet()) {
                    String resourceType = resourceEntry.getKey();
                    int amount = resourceEntry.getValue();
                    
                    availableResources.put(resourceType, availableResources.get(resourceType) + amount);
                }
                
                // Remove the task from completionTimes to avoid releasing resources multiple times
                completionTimes.remove(taskId);
                return; // We only release one task's resources at a time to avoid concurrent modification
            }
        }
    }
    
    /**
     * Selects the best task to execute next based on a priority heuristic.
     * 
     * Current implementation uses Earliest Deadline First with tie-breaking
     * on shortest execution time and then lexicographically by ID.
     */
    private Task selectBestTask(List<Task> eligibleTasks) {
        return Collections.min(eligibleTasks, (t1, t2) -> {
            // Priority 1: Earliest deadline first
            int deadlineComparison = Integer.compare(t1.getDeadline(), t2.getDeadline());
            if (deadlineComparison != 0) {
                return deadlineComparison;
            }
            
            // Priority 2: Shortest execution time
            int executionTimeComparison = Integer.compare(t1.getExecutionTime(), t2.getExecutionTime());
            if (executionTimeComparison != 0) {
                return executionTimeComparison;
            }
            
            // Priority 3: Lexicographical ordering of task IDs
            return t1.getId().compareTo(t2.getId());
        });
    }
    
    /**
     * Advances time to the next point where a task might become eligible.
     * This happens when resources are released after a task completes.
     */
    private boolean advanceTimeToNextEligibleTask(int currentTime, List<Task> tasks, 
                                                Set<String> scheduledTasks, 
                                                Map<String, Integer> completionTimes,
                                                Map<String, Integer> availableResources,
                                                Map<String, Integer> systemResources) {
        // Find the next time when resources will be released
        int nextReleaseTime = Integer.MAX_VALUE;
        for (int completionTime : completionTimes.values()) {
            if (completionTime > currentTime && completionTime < nextReleaseTime) {
                nextReleaseTime = completionTime;
            }
        }
        
        if (nextReleaseTime != Integer.MAX_VALUE) {
            // Advance time to the next release time
            int oldTime = currentTime;
            currentTime = nextReleaseTime;
            
            // Reset available resources to system resources
            for (Map.Entry<String, Integer> entry : systemResources.entrySet()) {
                availableResources.put(entry.getKey(), entry.getValue());
            }
            
            // Re-allocate resources for tasks that are still executing
            for (Map.Entry<String, Integer> entry : completionTimes.entrySet()) {
                String taskId = entry.getKey();
                int taskCompletionTime = entry.getValue();
                
                if (taskCompletionTime > currentTime) {
                    Task task = null;
                    for (Task t : tasks) {
                        if (t.getId().equals(taskId)) {
                            task = t;
                            break;
                        }
                    }
                    
                    if (task != null) {
                        allocateResources(task, availableResources);
                    }
                }
            }
            
            // Release resources for completed tasks
            releaseResources(currentTime, completionTimes, createTaskMap(tasks), availableResources);
            
            return true;
        }
        
        return false;
    }
    
    /**
     * Validates that the final schedule meets all constraints.
     */
    private boolean isValidSchedule(List<String> schedule, Map<String, Task> taskMap, 
                                   Map<String, Integer> systemResources) {
        if (schedule.isEmpty()) {
            return true;
        }
        
        Set<String> scheduledTasks = new HashSet<>();
        Map<String, Integer> availableResources = new HashMap<>(systemResources);
        int currentTime = 0;
        
        for (String taskId : schedule) {
            Task task = taskMap.get(taskId);
            
            // Check dependencies
            for (String dependency : task.getDependencies()) {
                if (!scheduledTasks.contains(dependency)) {
                    return false;
                }
            }
            
            // Check resources
            for (Map.Entry<String, Integer> entry : task.getResourceRequirements().entrySet()) {
                String resourceType = entry.getKey();
                int requiredAmount = entry.getValue();
                
                if (!availableResources.containsKey(resourceType) || 
                    availableResources.get(resourceType) < requiredAmount) {
                    return false;
                }
                
                // Allocate resources
                availableResources.put(resourceType, availableResources.get(resourceType) - requiredAmount);
            }
            
            // Update time
            currentTime += task.getExecutionTime();
            
            // Check deadline
            if (currentTime > task.getDeadline()) {
                return false;
            }
            
            // Add to scheduled tasks
            scheduledTasks.add(taskId);
            
            // Release resources (simplified for validation)
            for (Map.Entry<String, Integer> entry : task.getResourceRequirements().entrySet()) {
                String resourceType = entry.getKey();
                int requiredAmount = entry.getValue();
                
                availableResources.put(resourceType, availableResources.get(resourceType) + requiredAmount);
            }
        }
        
        return true;
    }
}

File: rust/exercises/practice/unit_test_6ebadf24_f22c_4079_8f08_ac64e46740f8/task_scheduler/tests/task_scheduler.rs
use task_scheduler::{Task, solve_task_scheduling};

#[test]
fn test_empty_tasks() {
    let tasks = vec![];
    assert_eq!(solve_task_scheduling(tasks), 0);
}

#[test]
fn test_single_task_no_dependency() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 5,
            deadline: 10,
            dependencies: vec![],
        }
    ];
    assert_eq!(solve_task_scheduling(tasks), 0); // No tardiness if deadline > processing time
}

#[test]
fn test_single_task_with_tardiness() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 10,
            deadline: 5,
            dependencies: vec![],
        }
    ];
    assert_eq!(solve_task_scheduling(tasks), 50); // (10-5)*10 = 50
}

#[test]
fn test_linear_dependency_chain() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 5,
            deadline: 5,
            dependencies: vec![],
        },
        Task {
            id: 1,
            processing_time: 3,
            deadline: 8,
            dependencies: vec![0],
        },
        Task {
            id: 2,
            processing_time: 4,
            deadline: 12,
            dependencies: vec![1],
        },
    ];
    assert_eq!(solve_task_scheduling(tasks), 0); // All tasks can be completed before deadlines
}

#[test]
fn test_linear_dependency_chain_with_tardiness() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 5,
            deadline: 3,
            dependencies: vec![],
        },
        Task {
            id: 1,
            processing_time: 3,
            deadline: 5,
            dependencies: vec![0],
        },
        Task {
            id: 2,
            processing_time: 4,
            deadline: 10,
            dependencies: vec![1],
        },
    ];
    // Task 0: Tardiness = 2, Weighted = 2*5 = 10
    // Task 1: Tardiness = 3, Weighted = 3*3 = 9
    // Task 2: Tardiness = 2, Weighted = 2*4 = 8
    // Total: 27
    assert_eq!(solve_task_scheduling(tasks), 27);
}

#[test]
fn test_circular_dependency() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 5,
            deadline: 10,
            dependencies: vec![1],
        },
        Task {
            id: 1,
            processing_time: 3,
            deadline: 8,
            dependencies: vec![0],
        },
    ];
    assert_eq!(solve_task_scheduling(tasks), -1); // Circular dependency, no valid schedule
}

#[test]
fn test_multiple_independent_tasks() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 5,
            deadline: 5,
            dependencies: vec![],
        },
        Task {
            id: 1,
            processing_time: 3,
            deadline: 3,
            dependencies: vec![],
        },
        Task {
            id: 2,
            processing_time: 4,
            deadline: 4,
            dependencies: vec![],
        },
    ];
    // Optimally scheduled: 1, 2, 0
    // Task 1: Completes at 3, no tardiness
    // Task 2: Completes at 7, tardiness = 3, weighted = 3*4 = 12
    // Task 0: Completes at 12, tardiness = 7, weighted = 7*5 = 35
    // Total: 47
    
    // But we can also do: 2, 1, 0 or other combinations
    // Let's check the best ordering
    
    // Order: 0, 1, 2
    // Task 0: Completes at 5, no tardiness
    // Task 1: Completes at 8, tardiness = 5, weighted = 5*3 = 15
    // Task 2: Completes at 12, tardiness = 8, weighted = 8*4 = 32
    // Total: 47
    
    // Order: 0, 2, 1
    // Task 0: Completes at 5, no tardiness
    // Task 2: Completes at 9, tardiness = 5, weighted = 5*4 = 20
    // Task 1: Completes at 12, tardiness = 9, weighted = 9*3 = 27
    // Total: 47
    
    // Order: 1, 0, 2
    // Task 1: Completes at 3, no tardiness
    // Task 0: Completes at 8, tardiness = 3, weighted = 3*5 = 15
    // Task 2: Completes at 12, tardiness = 8, weighted = 8*4 = 32
    // Total: 47
    
    // Order: 1, 2, 0
    // Task 1: Completes at 3, no tardiness
    // Task 2: Completes at 7, tardiness = 3, weighted = 3*4 = 12
    // Task 0: Completes at 12, tardiness = 7, weighted = 7*5 = 35
    // Total: 47
    
    // Order: 2, 0, 1
    // Task 2: Completes at 4, no tardiness
    // Task 0: Completes at 9, tardiness = 4, weighted = 4*5 = 20
    // Task 1: Completes at 12, tardiness = 9, weighted = 9*3 = 27
    // Total: 47
    
    // Order: 2, 1, 0
    // Task 2: Completes at 4, no tardiness
    // Task 1: Completes at 7, tardiness = 4, weighted = 4*3 = 12
    // Task 0: Completes at 12, tardiness = 7, weighted = 7*5 = 35
    // Total: 47
    
    // The minimum weighted tardiness is 47
    assert_eq!(solve_task_scheduling(tasks), 47);
}

#[test]
fn test_complex_dependency_graph() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 2,
            deadline: 5,
            dependencies: vec![],
        },
        Task {
            id: 1,
            processing_time: 3,
            deadline: 10,
            dependencies: vec![0],
        },
        Task {
            id: 2,
            processing_time: 4,
            deadline: 8,
            dependencies: vec![0],
        },
        Task {
            id: 3,
            processing_time: 2,
            deadline: 15,
            dependencies: vec![1, 2],
        },
        Task {
            id: 4,
            processing_time: 5,
            deadline: 20,
            dependencies: vec![3],
        },
    ];
    
    // Optimal schedule:
    // Task 0: 0-2, no tardiness
    // We need to decide task 1 or 2 next
    // If task 1: 2-5, no tardiness, then task 2: 5-9, tardiness = 1, weighted = 1*4 = 4
    // If task 2: 2-6, no tardiness, then task 1: 6-9, no tardiness
    // 
    // Optimal choice is task 2 then task 1:
    // Task 2: 2-6, no tardiness
    // Task 1: 6-9, no tardiness
    // Task 3: 9-11, no tardiness
    // Task 4: 11-16, no tardiness
    // Total: 0
    
    assert_eq!(solve_task_scheduling(tasks), 0);
}

#[test]
fn test_challenging_case() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 10,
            deadline: 20,
            dependencies: vec![],
        },
        Task {
            id: 1,
            processing_time: 15,
            deadline: 30,
            dependencies: vec![],
        },
        Task {
            id: 2,
            processing_time: 20,
            deadline: 40,
            dependencies: vec![0, 1],
        },
        Task {
            id: 3,
            processing_time: 5,
            deadline: 15,
            dependencies: vec![],
        },
        Task {
            id: 4,
            processing_time: 8,
            deadline: 25,
            dependencies: vec![3],
        },
        Task {
            id: 5,
            processing_time: 12,
            deadline: 35,
            dependencies: vec![2, 4],
        },
    ];
    
    // One possible optimal schedule:
    // Task 3: 0-5, no tardiness
    // Task 0: 5-15, no tardiness
    // Task 4: 15-23, no tardiness
    // Task 1: 23-38, tardiness = 8, weighted = 8*15 = 120
    // Task 2: 38-58, tardiness = 18, weighted = 18*20 = 360
    // Task 5: 58-70, tardiness = 35, weighted = 35*12 = 420
    // Total: 900
    
    // However, there may be a better ordering, so just testing that the result is <= 900
    let result = solve_task_scheduling(tasks);
    assert!(result <= 900 && result > 0);
}

#[test]
fn test_medium_sized_problem() {
    let tasks = vec![
        Task { id: 0, processing_time: 5, deadline: 10, dependencies: vec![] },
        Task { id: 1, processing_time: 3, deadline: 8, dependencies: vec![0] },
        Task { id: 2, processing_time: 7, deadline: 15, dependencies: vec![1] },
        Task { id: 3, processing_time: 2, deadline: 12, dependencies: vec![0] },
        Task { id: 4, processing_time: 4, deadline: 20, dependencies: vec![2, 3] },
        Task { id: 5, processing_time: 6, deadline: 25, dependencies: vec![4] },
        Task { id: 6, processing_time: 8, deadline: 30, dependencies: vec![5] },
        Task { id: 7, processing_time: 3, deadline: 18, dependencies: vec![3] },
        Task { id: 8, processing_time: 5, deadline: 22, dependencies: vec![7] },
        Task { id: 9, processing_time: 4, deadline: 28, dependencies: vec![8, 6] },
    ];
    
    // This is a moderately complex graph with multiple paths
    // Just testing that the function returns a valid result (not -1)
    let result = solve_task_scheduling(tasks);
    assert!(result >= 0);
}

#[test]
fn test_multiple_possible_paths() {
    let tasks = vec![
        Task {
            id: 0,
            processing_time: 2,
            deadline: 5,
            dependencies: vec![],
        },
        Task {
            id: 1,
            processing_time: 3,
            deadline: 7,
            dependencies: vec![],
        },
        Task {
            id: 2,
            processing_time: 4,
            deadline: 10,
            dependencies: vec![0, 1],
        },
        Task {
            id: 3,
            processing_time: 5,
            deadline: 12,
            dependencies: vec![0],
        },
        Task {
            id: 4,
            processing_time: 3,
            deadline: 15,
            dependencies: vec![2, 3],
        },
    ];
    
    // We have multiple valid paths through this dependency graph
    // Testing that we get a valid result
    let result = solve_task_scheduling(tasks);
    assert!(result >= 0);
}